### ISSUE TITLE
`cargo upgrade` should honour how specific a dependency is

### ISSUE THREAD

**Issue by FauxFaux** [rust-lang/cargo#4727]
If my `Cargo.toml` contains:

```toml
[dependencies]
nom = "3"
```

then `cargo upgrade` will make the version very explicit, changing it to "3.2.1".
```
$ cargo upgrade
Upgrading nom v3 -> v3.2.1
```

This (as I understand it) has changed the dependency specification from `(>=3.0.0, <4.0.0)` to `(>=3.2.1, <3.3.0)`. This means that `cargo` will have a harder time resolving dependencies in future, i.e. if another library ran `cargo upgrade` at a different time and randomly selected `3.3.7` `(>=3.3.7, <3.4.0)` as its specification.

I would expect `cargo upgrade` to increase only visible version numbers, so `nom = "2"` -> `nom = "3"` would be okay, but to leave `nom = "3"` alone.

Currently, `0.5` will be changed to `0.5.1` (which I feel is wrong), and it should be left at `0.5`, or changed to `0.6` (if available).

What if `9.2.6` is available for this library which is currently at `7.5`? I feel that it could be changed to `9`.

Thoughts?

**Comment by sinkuu** [comment]
`cargo upgrade` never modifies `Cargo.toml`. It just recorded in `Cargo.lock` the latest available `nom` version (`3.2.1`) according to your specificaiton (`3`).

**Comment by FauxFaux** [comment]
Oh dear. I thought `cargo upgrade` (not `cargo update` which @sinkuu is thinking of) was a built-in. It's not, it's part of https://github.com/killercup/cargo-edit . My bad!
